core/src/main/java/io/grpc/transport/AbstractStream.java

```diff
@@ -351,28 +321,25 @@ protected final void onSendingBytes(int numBytes) {
   protected final void onSentBytes(int numBytes) {
     boolean doNotify;
     synchronized (onReadyLock) {
+      boolean belowThresholdBefore = numSentBytesQueued < onReadyThreshold;
       numSentBytesQueued -= numBytes;
-      doNotify = needToNotifyOnReady();
+      boolean belowThresholdAfter = numSentBytesQueued < onReadyThreshold;
+      doNotify = !belowThresholdBefore && belowThresholdAfter;
     }
     if (doNotify) {
-      listener().onReady();
+      notifyIfReady();
     }
   }
 
-  /**
-   * Determines whether or not we need to call the {@link StreamListener#onReady()} handler now.
-   * Calling this method has the side-effect of unsetting {@link #shouldNotifyOnReady} so the
-   * handler should always be invoked immediately after calling this method.
-   */
-  @GuardedBy("onReadyLock")
-  private boolean needToNotifyOnReady() {
-    if (shouldNotifyOnReady && isReady()) {
-      // Returning true here counts as a call to the onReady callback, so
-      // unset the flag.
-      shouldNotifyOnReady = false;
-      return true;
+  @VisibleForTesting
+  final void notifyIfReady() {
```

Comment:
Acquiring the lock only once greatly eases proving correctness. I don't see the benefit of only calling `onReady()` from one place outweighing the greater difficulty in auditing. It seems the "simple function" doesn't really help the methods that are calling it, in that they would be more clear and the overall code shorter if they just did the logic inline.

It _can_ go in as-is, but I do have reservations about it.